ble/CallChainOfFunctionPointersWithContext.h

```diff
@@ -120,11 +161,40 @@ class CallChainOfFunctionPointersWithContext {
      *        chained FunctionPointers.
      */
     void call(ContextType context) {
-        if (chainHead) {
-            chainHead->call(context);
+        ((const CallChainOfFunctionPointersWithContext*) this)->call(context);
+    }
+
+    /**
+     * @brief same as above but const 
+     */
+    void call(ContextType context) const {
+        currentCalled = chainHead;
+
+        while(currentCalled) { 
+            currentCalled->call(context);
+            // if this was the head and the call removed the head
+            if(currentCalled == NULL) { 
+                currentCalled = chainHead;
+            } else {
+                currentCalled = currentCalled->getNext();
+            }
         }
     }
 
+    /**
+     * @brief same as above but with function call operator
+     */
+    void operator()(ContextType context) const {
+        call(context);
+    }
+
+    typedef void (CallChainOfFunctionPointersWithContext::*bool_type)() const;
+    void True() const {}
+
+    operator bool_type() const {
```

Comment:
operator bool open a hole in the type system, it allow the comparison of unrelated types if they implement this operator. 

```c++
struct A { 
    operator bool() { return true; }
};

struct B { 
    operator bool() { return true; } 
}

A a;
B b;

// this will compile!!!
if(a == b) { 
// we go there  ....
}
```

I will make an abstraction to generate safe_bool class, it will be more understandable